### PR TITLE
[RLlib; Off-policy] Add episode sampling to `EpisodeReplayBuffer`.

### PR DESCRIPTION
**Pull request by simonsays1980**
## Why are these changes needed?

The `EpisodeReplayBuffer` is still sampling batches, i.e. dicts. This PR proposes a way to add episode sampling to the buffer such that the buffer can be used in the same form as the other episode buffers. This is also a preliminary for using buffers in the `OfflinePreLearner`.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/utils/replay_buffers/episode_replay_buffer.py [hunk 1]
```diff
@@ -332,6 +414,179 @@ def sample(
...
+        *,
+        batch_size_B: Optional[int] = None,
+        batch_length_T: Optional[int] = None,
+        n_step: Optional[Union[int, Tuple]] = None,
+        gamma: float = 0.99,
+        include_infos: bool = False,
+        include_extra_model_outputs: bool = False,
+        **kwargs,
+    ) -> List[SingleAgentEpisode]:
+        """Samples episodes from a buffer in a randomized way.
```
Comment (sven1977) [inline rllib/utils/replay_buffers/episode_replay_buffer.py:429]: I think this docstrings is not accurate ??

File: rllib/utils/replay_buffers/episode_replay_buffer.py [hunk 2]
```diff
@@ -211,6 +212,87 @@ def sample(
...
+                If `n_step` is a tuple, it is considered as a range to sample
+                from. If `None`, we use `n_step=1`.
+            gamma: The discount factor to be used when applying n-step calculations.
+                The default of `0.99` should be replaced by the `Algorithm`s
+                discount factor.
+            include_infos: A boolean indicating, if `info`s should be included in
+                the batch. This could be of advantage, if the `info` contains
+                values from the environment important for loss computation. If
+                `True`, the info at the `"new_obs"` in the batch is included.
+            include_extra_model_outputs: A boolean indicating, if
```
Comment (sven1977) [inline rllib/utils/replay_buffers/episode_replay_buffer.py:258]: add `sample_episodes` arg description here

File: rllib/utils/replay_buffers/episode_replay_buffer.py [hunk 3]
```diff
@@ -211,6 +212,87 @@ def sample(
...
         batch_length_T: Optional[int] = None,
+        n_step: Optional[Union[int, Tuple]] = None,
+        beta: float = 0.0,
+        gamma: float = 0.99,
+        include_infos: bool = False,
+        include_extra_model_outputs: bool = False,
+        sample_episodes: Optional[bool] = False,
+        **kwargs,
+    ) -> Union[SampleBatchType, SingleAgentEpisode]:
+        """Samples from a buffer in a randomized way.
```
Comment (sven1977) [inline rllib/utils/replay_buffers/episode_replay_buffer.py:223]: update docstring to reflect what the behavior is for `sample_episodes`.

**Comment by sven1977** [comment]
Awesome unification PR! This opens the door to a DQfD style mixing of offline and off-policy RL!

Dumb question: Would this already work even if the PrioritizedEpisodeReplayBuffer doesn't support this flag yet?

**Comment by simonsays1980** [comment]
> Awesome unification PR! This opens the door to a DQfD style mixing of offline and off-policy RL!
> 
> Dumb question: Would this already work even if the PrioritizedEpisodeReplayBuffer doesn't support this flag yet?

Thanks @sven1977! I guess that mixing would be possible even before? Wouldn't it? We just use DQN and add a dataset, then sample from the buffer and let the `OfflinePreLearner` return episodes instead of `MultiAgentBatch` to be able to mix the two episode lists. Then pass the lists to the trainer. We cannot make use of the offline pipeline because we need to add continuously episodes to the off-policy buffer and this one should exist over the full course of training.

The `PrioritizedEpisodeReplayBuffer` did already support the episode sampling and could be used in the `OfflinePreLearner`. Imo this, however, would somehow bias offline learning because we would filter data between offline data and learner. But, if a user wants to have a prioritized sampling from her offline data, we could support this out-of-the-box :) Will add a buffer definition to the `AlgorithmConfig` of `offline()`.

**Review by sven1977 — APPROVED** [review]
LGTM now. Thanks a ton @simonsays1980 !!

**Pull request merged**